pydicom/pixel_data_handlers/util.py

```diff
@@ -402,34 +414,26 @@ def _convert_RGB_to_YBR_FULL(arr):
 
     * DICOM Standard, Part 3,
       :dcm:`Annex C.7.6.3.1.2<part03/sect_C.7.6.3.html#sect_C.7.6.3.1.2>`
-    * ISO/IEC 10918-5:2012, Section 7
+    * ISO/IEC 10918-5:2012 (`ITU T.871
+      <https://www.ijg.org/files/T-REC-T.871-201105-I!!PDF-E.pdf>`_),
+      Section 7
     """
     orig_dtype = arr.dtype
-    arr = arr.astype(np.float)
 
     rgb_to_ybr = np.asarray(
-        [[+0.299, +0.587, +0.114],
-         [-0.299, -0.587, +0.886],
-         [+0.701, -0.587, -0.114]], dtype=np.float)
-
-    arr = np.dot(arr, rgb_to_ybr.T)
-    if len(arr.shape) == 4:
-        # Multi-frame
-        arr[:, :, :, 1] /= 1.772
-        arr[:, :, :, 2] /= 1.402
-    else:
-        # Single frame
-        arr[:, :, 1] /= 1.772
-        arr[:, :, 2] /= 1.402
-
-    arr += [0, 128, 128]
-
-    # Round(x) -> floor of (arr + 0.5)
-    arr = np.floor(arr + 0.5)
+        [[+0.299, -0.299 / 1.772, +0.701 / 1.402],
+         [+0.587, -0.587 / 1.772, -0.587 / 1.402],
+         [+0.114, +0.886 / 1.772, -0.114 / 1.402]],
+        dtype=np.float
```

Comment:
I'm having trouble to understand the change here - the previous matrix has been taken right from the standard, and I understand that you are doing some optimization here, but I don't understand the rearrangement of the matrix components (matrix manipulations have never been my strong side, and I learned that a long time ago, so bear with me...). Can you elaborate a bit, please?

Reply:
The T.871 formula without the clipping is:
```
Y  =   0.299 * R + 0.587 * G + 0.114 * B  / 1.000 +   0
Cb = (-0.299 * R - 0.587 * G + 0.886 * B) / 1.772 + 128
Cr = ( 0.701 * R - 0.587 * G - 0.114 * B) / 1.402 + 128
```
Which is the same as the following (for a single pixel):
```
A = [R, G, B]  # = ds.pixel_array[0, 0, :] for single frame data
# This is the current `rgb_to_ybr`
B = [
    [+0.299, +0.587, +0.114],
    [-0.299, -0.587, +0.886],
    [+0.701, -0.587, -0.114]
]

# Then...
[Y, Cb, Cr] = A dot B^T

# Where B^T is the transpose of B:
B_T = [
    [ 0.299, -0.299,  0.701],
    [ 0.587, -0.587, -0.587],
    [ 0.114,  0.886, -0.114]
]

# And:
A dot B_T = [
    [  R * 0.299 + G * 0.587 + B * 0.114],  # Y, RGB times the 1st column of B_T
    [- R * 0.299 - G * 0.587 + B * 0.886],  # Cb, RGB times the 2nd column of B_T
    [  R * 0.701 - G * 0.587 - B * 0.114]   # Cr, RGB times the 3rd column of B_T
]

# And dividing and adding to the Cb and Cr channels 
Y  =  Y / 1 + 0  # No change
Cb = Cb / 1.772 + 128
Cr = Cr / 1.402 + 128
```

Which is how the original works:
```python
rgb_to_ybr = np.asarray(
    [[+0.299, +0.587, +0.114],
     [-0.299, -0.587, +0.886],
     [+0.701, -0.587, -0.114]], dtype=np.float)

arr = np.dot(arr, rgb_to_ybr.T)
if len(arr.shape) == 4:
    # Multi-frame
    arr[:, :, :, 1] /= 1.772
    arr[:, :, :, 2] /= 1.402
else:
    # Single frame
    arr[:, :, 1] /= 1.772
    arr[:, :, 2] /= 1.402

arr += [0, 128, 128]
```

However its simpler (and quicker due to fewer array operations) to define `rgb_to_ybr` so that its already the [transpose ](https://en.wikipedia.org/wiki/Transpose)and already divides the Cb and Cr channels
```
# Only the transposition
rgb_to_ybr = [
    [ 0.299, -0.299,  0.701],
    [ 0.587, -0.587, -0.587],
    [ 0.114,  0.886, -0.114]
]
# Tranposition + dividing what will become the Cb and Cr channels
# We divide the 2nd and 3rd columns as these correspond to what
#   will become the Cb and Cr channels, respectively (see above)
rgb_to_ybr = [
    [ 0.299, -0.299 / 1.772,  0.701 / 1.402],
    [ 0.587, -0.587 / 1.772, -0.587 / 1.402],
    [ 0.114,  0.886 / 1.772, -0.114 / 1.402]
]
# We can do this because:
Cb = (- R * 0.299 + G * 0.587 + B * 0.886) / 1.772
# Is equivalent to:
Cb = - R * 0.299 / 1.772 + G * 0.587 / 1.772 + B * 0.886 / 1.772
```
With the added bonus that the `if else` statement for multi-frame data is no longer
necessary.